version: '4'

services:
  zookeeper:
    image: wurstmeister/zookeeper:latest
    ports:
      - "2181:2181"

  kafka:
    image: wurstmeister/kafka:latest
    ports:
      - "9092:9092"
    environment:
      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka:9092
      KAFKA_LISTENERS: PLAINTEXT://0.0.0.0:9092
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_LOG4J_LOGGERS: "kafka=ERROR, state.change.logger=ERROR"
    depends_on:
      - zookeeper

  spark:
    image: bitnami/spark:latest
    environment:
      SPARK_MODE: master
    ports:
      - "8080:8080"

  producer:
    build:
      context: ./producer
    volumes:
      - ./data:/data
    depends_on:
      - kafka
    command: python data_producer.py

  consumer:
    build:
      context: ./consumer
    depends_on:
      - kafka
      - spark
      - postgres
    command: python data_consumer.py 2> error_log.log

  postgres:
    build:
      context: ./postgres
    environment:
      POSTGRES_DB: mydb
      POSTGRES_USER: user
      POSTGRES_PASSWORD: password
    ports:
      - "5432:5432"
    volumes:
      - ./postgres:/docker-entrypoint-initdb.d

  ml_app:
    build:
      context: ./ml_app
    ports:
      - "8501:8501"
    depends_on:
      - postgres
      - spark
    command: /bin/bash -c "streamlit run main.py > logs/output.log 2>&1"

